### ISSUE TITLE
Consider whether spanner::Value should support an "is_empty()" member

### ISSUE THREAD

**Issue by devjgm** [googleapis/google-cloud-cpp-spanner#157]
In #156 I'm adding a default constructor, which creates an "empty" Value with no type. The question was raised as to whether or not we should add an `is_empty()` member to Value.

I think the semantics would be: `is_empty()` returns true IFF the `Value` holds no type or value information. That is, whether or not any contained type or value information that may exist is *valid* is not verified.

In other words, a default constructed Value would be "empty". Any other value would not be.

Do we want to add this support?

I'd be curious to hear @devbww 's thoughts.

**Comment by coryan** [comment]
I think the semantics should be:

"`is_empty()` returns true IFF there is not type **nor** value information."

Though I guess if you have value information then you necessarily do have type information too.  Boolean tables, how do they work? :confused:

**Comment by devbww** [comment]
`Value` should maintain the invariant "has_type\<T\> => has_value\<T\>" (where "cannot happen" encoding errors don't refute "has_value\<T\>").  So, the **or/nor** question wouldn't arise.
```
Value::MakeTypeProto() {
  google::spanner::v1::Type t;
  t.set_code(google::spanner::v1::TypeCode::TYPE_CODE_UNSPECIFIED);
  return t;
}

Value::Value() : type_(MakeTypeProto()) {}

// For all T: !is<T>().
bool Value::is_empty() {
  return type_ == MakeTypeProto();
}
```

As to whether `is_empty()` should even exist, how do we envisage a user needing it?
```
Value v;
// Some code that conditional resets "v".
if (!v.is_empty()) {
   // Use "v" to make a query/mutation/etc.
}
```
So, we're saving them from having to keep that state themselves?  OK.

In general I would say that if the default ctor exists, then yes, there should be a way to check for that value.  If we further specify the default ctor ...
```
Value::Value() : type_(MakeTypeProto()), value_(MakeValueProto()) {}
```
..., then `Value` equality would suffice for that check.
```
Value v;
// Some code that conditional resets "v".
if (v != Value()) {
   // Use "v" to make a query/mutation/etc.
}
```

**Comment by coryan** [comment]
`if (v != Value())` sounds better than a `.is_empty()` member function. *Provided*  we fix the documentation in:

https://github.com/googleapis/google-cloud-cpp-spanner/blob/5c38c25f74d1ca8e44c7998edccc04f7f13cca11/google/cloud/spanner/value.h#L143

**Comment by devjgm** [comment]
I *think* the `if (v != Value())` check would work now. I don't see what would need to change w/ the documentation about the default c'tor.

**Comment by coryan** [comment]
The current doc can be paraphrased as "These are all the things that will return `false` when you try to determine if a `Value` was initialized with this constructor ... we are not telling you what will return `true`". 
 You are leaving it to the user to figure out that they can call `v == Value()` to get something that returns `true` only if a `v` value was initialized with the default constructor.

**Comment by devjgm** [comment]
Oh, I see. Thanks. I guess I thought it was obvious that `Value() == Value()`, so I documented the less obvious things. But sure, I'm totally open to any and all doc improvements.
